src/os.c

```diff
@@ -240,18 +262,30 @@ void  _mi_os_free(void* p, size_t size, mi_stats_t* stats) {
 // address that should be available then.
 static void* mi_os_alloc_aligned_ensured(size_t size, size_t alignment, size_t trie, mi_stats_t* stats)
 {
+  // allocate a chunk that includes the alignment
+#ifdef __wasi__
+  uintptr_t page_size = _mi_os_page_size();
+  uintptr_t base = __builtin_wasm_memory_size(0) * page_size;
+  uintptr_t aligned_base = _mi_align_up(base, (uintptr_t) alignment);
+  size_t alloc_size = aligned_base - base + size;
+  mi_assert(alloc_size >= size);
+  if (alloc_size < size) return NULL;
+  if (__builtin_wasm_memory_grow(0, alloc_size / page_size) == SIZE_MAX) {
+    errno = ENOMEM;
+    return NULL;
+  }
+  void* aligned_p = (void*) aligned_base;;
+#else
   if (trie >= 3) return NULL; // stop recursion (only on Windows)
   size_t alloc_size = size + alignment;
   mi_assert(alloc_size >= size); // overflow?
   if (alloc_size < size) return NULL;
-
-  // allocate a chunk that includes the alignment
   void* p = mi_mmap(NULL, alloc_size, 0, stats);
   if (p == NULL) return NULL;
   // create an aligned pointer in the allocated area
   void* aligned_p = mi_align_up_ptr(p, alignment);
   mi_assert(aligned_p != NULL);
-#if defined(_WIN32)
+  #if defined(_WIN32)
   // free it and try to allocate `size` at exactly `aligned_p`
```

Comment:
There's still a whitespace change here.

Reply:
Ah, I see that it's an intentional indentation... sorry for the noise.

Reply:
Trying to match the existing indentation style.